### PR TITLE
Some fixes to finality tests

### PR DESCRIPTION
**Pull request by djrtwo**
Fix finality_12. This should be a good base to fix the others.

* need to set the associated epoch boundary block roots to ensure that `get_matching_target_` works
* justification bit 0 should be set at first so it gets shift to position 1. Then 0 gets justified and you have 0 and 1 for finality.
* `aggregation_bitfield` -> `aggregation_bits` in `add_mock_attestations` (use bool array for init)
* `add_mock_attestations` had bug, setting `target=source`. fixed
* `add_mock_attestations` sets `shard` in `attestation.data.crosslink`. This piece of data is required to figure out what committee is attesting. Previous, all were defaulting to zero so each committee added looked the same.

### CODE REVIEW

File: test_libs/pyspec/eth2spec/test/phase_0/epoch_processing/test_process_justification_and_finalization.py [hunk 1]
```diff
@@ -147,16 +149,19 @@ def finalize_on_123(spec, state, epoch, support):
...
     c3 = spec.Checkpoint(epoch=epoch - 3, root=b'\xaa' * 32)
     c2 = spec.Checkpoint(epoch=epoch - 2, root=b'\xbb' * 32)
     c1 = spec.Checkpoint(epoch=epoch - 1, root=b'\xcc' * 32)
+    state.block_roots[spec.get_epoch_start_slot(c3.epoch) % spec.SLOTS_PER_HISTORICAL_ROOT] = c3.root
+    state.block_roots[spec.get_epoch_start_slot(c2.epoch) % spec.SLOTS_PER_HISTORICAL_ROOT] = c2.root
+    state.block_roots[spec.get_epoch_start_slot(c1.epoch) % spec.SLOTS_PER_HISTORICAL_ROOT] = c1.root
 
     old_finalized = state.finalized_checkpoint
-    state.previous_justified_checkpoint = c3
+    state.previous_justified_checkpoint = c4
```
Comment (CarlBeek) [inline test_libs/pyspec/eth2spec/test/phase_0/epoch_processing/test_process_justification_and_finalization.py]: It feels strange to do this as it is not the behaviour one would expect from this test. The problem, however, is that otherwise, a new checkpoint is finalised as the condition of "2nd/3rd most recent epochs are justified, the 2nd using the 3rd as source" is met. This means that in the case where there is insufficient support, `assert state.finalized_checkpoint == old_finalized  # no new finalized` fails. @djrtwo, please can you check me here (and propose an alternative solution if you have one.)

TLDR: this is required because of an artifact of how the tests are constructed by manipulating what is and isn't justified in the state.

**Comment by CarlBeek** [comment]
I extended these finalisation fixes to all of the finalisation cases. Fixes include:
* all of the `boundry_epochs` are now set (needed for the correct calculation of attestations).
* Many list-slicing bugs
* correction of what is justified in each of the tests.

**Comment by protolambda** [comment]
Fixed the attestation creation, since it was broken on mainnet config (multi committees per slot). And also reworded the bitfield indices and justification, it's clear now.
